Implement trusted types integrations with DOM attribute APIs.

ziransun opened the pull request:
#### 9a38b690ee9a1494dc87e6cbb58c3ec7ce9cda9b
<pre>
Implement trusted types integrations with DOM attribute APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270436">https://bugs.webkit.org/show_bug.cgi?id=270436</a>

Reviewed by Ryosuke Niwa.

Implement the spec updates at <a href="https://github.com/whatwg/dom/pull/1247">https://github.com/whatwg/dom/pull/1247</a>

It also removes some expectations in GTK as the results should be
in line with the general expectation file.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-setAttribute-respects-Elements-node-documents-globals-CSP-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/GlobalEventHandlers-onclick-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttributeNS-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/dom/Element.cpp:
(WebCore::trustedTypesCompliantAttributeValue):
(WebCore::Element::validateAttributeIndex const):
(WebCore::Element::toggleAttribute):
(WebCore::Element::setAttribute):
(WebCore::Element::setElementsArrayAttribute):
(WebCore::appendAttributes):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::setAttributeNS):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/dom/TrustedScript.h:
* Source/WebCore/dom/TrustedScriptURL.h:
(WebCore::TrustedScriptURL::toString const): Deleted.
(WebCore::TrustedScriptURL::toJSON const): Deleted.
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::stringToTrustedType):
(WebCore::trustedTypeForAttribute):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
(WebCore::TrustedTypePolicyFactory::getAttributeType const):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMElement.mm:
(-[WKDOMElement setAttribute:value:]):
* Source/WebKitLegacy/mac/DOM/DOMElement.mm:
(-[DOMElement setAttribute:value:]):
(-[DOMElement setAttributeNS:qualifiedName:value:]):

Canonical link: <a href="https://commits.webkit.org/278817@main">https://commits.webkit.org/278817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04b9b3726c091ba5fac6c70fce7d5bcbb5382c4d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54882 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2308 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53919 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37277 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1990 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42016 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53715 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28570 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44524 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23144 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25873 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1783 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47833 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1873 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56474 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1748 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49415 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27974 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44593 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48627 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11304 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28871 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/dom/Element.h:
```diff
@@ -35,6 +35,7 @@
 #include "ScrollTypes.h"
 #include "ShadowRootMode.h"
 #include "SimulatedClickOptions.h"
+#include "TrustedType.h"
```
I don't know that you can for sure do it, but you should try to use a forward declaration instead.

just doing in the block around line 50.
```
class TrustedHTML;
class TrustedScript;
class TrustedScriptURL;
```

lukewarlow replied:
If you can't do a forward declaration you at the very least wanna change this to just import the type headers directly. And import TrustedType.h inside of Element.cpp

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -81,6 +82,19 @@ ASCIILiteral trustedTypeToString(TrustedType trustedType)
 
+TrustedType stringToTrustedType(String str)
+{
+    if (str == "TrustedHTML"_s)
```
Probably could use switch instead?

lukewarlow commented on Source/WebCore/dom/TrustedType.h:
```diff
@@ -50,4 +51,5 @@ WEBCORE_EXPORT ExceptionOr<String> trustedTypeCompliantString(TrustedType, Scrip
 
 WEBCORE_EXPORT ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecutionContext&, const String& urlString);
 
+WEBCORE_EXPORT String getAttributeTrustedType(const String& tagName, const String& attributeParameter, const String& elementNamespace, const String& attributeNamespace, String& sink);
```
nit: getTrustedTypeForAttribute?

Not sure if that's better or worse just an idea.

ziransun replied:
Done. However, we still have the build error in Mac. Not sure if this is related.

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1984,7 +2008,15 @@ ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const A
     auto name = index != ElementData::attributeNotFound ? attributeAt(index).name() : QualifiedName { nullAtom(), caseAdjustedQualifiedName, nullAtom() };
-    setAttributeInternal(index, name, value, InSynchronizationOfLazyAttribute::No);
+
+    auto sink = nullString();
```
Worth adding a flag check here to skip all the trusted types related code if the flag is off. The code does handle it itself but this will be a hotpath so no harm in doing it early.

ziransun replied:
It's handled at L1983, so not sure we need the flag check here.

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2288,7 +2320,7 @@ void Element::setElementsArrayAttribute(const QualifiedName& attributeName, std:
     auto newElements = copyToVectorOf<WeakPtr<Element, WeakPtrImplWithEventTargetData>>(*elements);
     explicitlySetAttrElementsMap().set(attributeName, WTFMove(newElements));
-    
+
```
Probably worth reverting this whitespace only change?

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3195,7 +3227,7 @@ static void appendAttributes(StringBuilder& builder, const Element& element)
             addEllipsis = true;
         }
-        
+
```
ditto?

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -170,6 +184,48 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
+    auto attributeName = attributeParameter.convertToASCIILowercase();
+
+    if (attributeName.startsWith("on"_s))
+        return trustedTypeToString(TrustedType::TrustedScript);
```
This path should set the sink value too. Not sure how best to calculate the value for this.

lukewarlow replied:
Actually scratch that I don't think it needs to because it can only be hit via setAttribute. Rather than the IDL attribute path.

lukewarlow commented on Source/WebCore/dom/TrustedType.h:
```diff
@@ -50,4 +51,5 @@ WEBCORE_EXPORT ExceptionOr<String> trustedTypeCompliantString(TrustedType, Scrip
 
 WEBCORE_EXPORT ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecutionContext&, const String& urlString);
 
+WEBCORE_EXPORT String getAttributeTrustedType(const String& tagName, const String& attributeParameter, const String& elementNamespace, const String& attributeNamespace, String& sink);
```
maybe init sink to be a nullString() here?

lukewarlow commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -103,30 +103,8 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
-
-    return nullString();
+    auto sink = nullString();
+    return getAttributeTrustedType(tagName, attributeParameter, elementNamespace, attributeNamespace, sink);
```
If you init sink to be a nullString in the header then you wont need it here.

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3363,12 +3395,20 @@ ExceptionOr<QualifiedName> Element::parseAttributeName(const AtomString& namespa
         return result.releaseException();
-    setAttribute(result.releaseReturnValue(), value);
+
+    auto sink = nullString();
```
ditto wrt flag check

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -170,6 +184,48 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
+        return trustedTypeToString(TrustedType::TrustedHTML);
+    }
+    if (element.matches(HTMLNames::embedTag) && attribute.matches(HTMLNames::srcAttr)) {
+        sink = "HTMLEmbedElement src"_s;
```
For now can we remove the sink parameter and logic from this function please.

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1984,7 +2008,15 @@ ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const A
-    setAttributeInternal(index, name, value, InSynchronizationOfLazyAttribute::No);
+
+    auto sink = nullString();
+    String attributeType = getAttributeTrustedType(name.localName(), getAttribute(name), ""_s, ""_s, sink);
```
Remove sink param from this function and remove the variable

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1984,7 +2008,15 @@ ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const A
+
+    auto sink = nullString();
+    String attributeType = getAttributeTrustedType(name.localName(), getAttribute(name), ""_s, ""_s, sink);
+    auto attributeValue = getTrustedTypesCompliantAttributeValue(attributeType, value, this, sink);
```
Change this sink param to just be "Element setAttribute"_s

I know this doesn't match the specification but the spec is wrong (https://github.com/w3c/trusted-types/issues/492)

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3363,12 +3395,20 @@ ExceptionOr<QualifiedName> Element::parseAttributeName(const AtomString& namespa
+
+    auto sink = nullString();
+    String attributeType = getAttributeTrustedType(qualifiedName, getAttribute(qualifiedName), ""_s, namespaceURI, sink);
+    auto attributeValue = getTrustedTypesCompliantAttributeValue(attributeType, value, this, sink);
```
This sink should be "Element setAttributeNS"

lukewarlow commented on Source/WebCore/dom/Element.idl:
```diff
@@ -42,8 +42,8 @@
     [ImplementedAs=getAttributeNSForBindings] DOMString? getAttributeNS([AtomString] DOMString? namespaceURI, [AtomString] DOMString localName);
-    [CEReactions=Needed] undefined setAttribute([AtomString] DOMString qualifiedName, [AtomString] DOMString value);
-    [CEReactions=Needed] undefined setAttributeNS([AtomString] DOMString? namespaceURI, [AtomString] DOMString qualifiedName, [AtomString] DOMString value);
+    [CEReactions=Needed] undefined setAttribute([AtomString] DOMString qualifiedName, (TrustedType or DOMString) value);
```
You should keep the [AtomString] annotation on the DOMString type. (I think anyway)

lukewarlow replied:
Be aware that will require changing some of the C++

lukewarlow commented on Source/WebCore/dom/Element.idl:
```diff
@@ -42,8 +42,8 @@
-    [CEReactions=Needed] undefined setAttribute([AtomString] DOMString qualifiedName, [AtomString] DOMString value);
-    [CEReactions=Needed] undefined setAttributeNS([AtomString] DOMString? namespaceURI, [AtomString] DOMString qualifiedName, [AtomString] DOMString value);
+    [CEReactions=Needed] undefined setAttribute([AtomString] DOMString qualifiedName, (TrustedType or DOMString) value);
+    [CEReactions=Needed] undefined setAttributeNS([AtomString] DOMString? namespaceURI, [AtomString] DOMString qualifiedName, (TrustedType or DOMString) value);
```
ditto

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1983,8 +2008,18 @@ ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const A
+        if (attributeValue.hasException())
+            return attributeValue.releaseException();
+
+        setAttributeInternal(index, name,  AtomString(attributeValue.releaseReturnValue()), InSynchronizationOfLazyAttribute::No);
```
This index can be stale as `getTrustedTypesCompliantAttributeValue` can result in JS execution which may mutate the attributes of the element. We need to work out an interopable solution to this.

ziransun replied:
As discussed, we will look this at a later stage. The bug is filed at https://bugs.webkit.org/show_bug.cgi?id=272678

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -170,6 +184,36 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
+    auto localName = tagName.convertToASCIILowercase();
+    auto attributeName = attributeParameter.convertToASCIILowercase();
+
+    if (attributeName.startsWith("on"_s))
```
The spec for this has been updated to use the "event handler content attribute" concept. If you can easily match the attributeName to event handler names then great. If not feel free to leave this for now and we can address in a follow up.

lukewarlow replied:
https://pr-preview.s3.amazonaws.com/lukewarlow/trusted-types/pull/457.html#get-trusted-type-data-for-attribute - latest preview of spec

lukewarlow replied:
Dw about the sink values we can addres that in a follow up because the tests probably need updating too.

ziransun replied:
A follow-up bug is filed at https://bugs.webkit.org/show_bug.cgi?id=272675

ziransun replied:
Code updated to be in line with the spec update. I will remove bug 272675 once this is landed.

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -170,6 +184,36 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
+        return trustedTypeToString(TrustedType::TrustedScriptURL);
+    if (element.matches(HTMLNames::scriptTag) && (attribute.matches(HTMLNames::srcAttr)))
+        return trustedTypeToString(TrustedType::TrustedScriptURL);
+    if (element.matches(HTMLNames::objectTag) && attribute.matches(HTMLNames::dataAttr))
```
The embed and object protection has been removed so you need to rebase this branch and then make sure there's no objectTag or embedTag references reintroduced.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2000,7 +2001,31 @@ ExceptionOr<bool> Element::toggleAttribute(const AtomString& qualifiedName, std:
 }
 
-ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const AtomString& value)
+static ExceptionOr<String> getTrustedTypesCompliantAttributeValue(const String& attributeType, const TrustedTypeOrString& value, Element* element, String sink)
```
"get" prefix should only be used for a function with an out argument.
This should be named `trustedTypesCompliantAttributeValue` instead.

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1975,6 +1976,43 @@ const AtomString& Element::getAttributeNS(const AtomString& namespaceURI, const
     return getAttribute(QualifiedName(nullAtom(), localName, namespaceURI));
 }
 
+static ExceptionOr<String> trustedTypesCompliantAttributeValue(const String& attributeType, const TrustedTypeOrString& value, Element* element, String sink)
```
Change this to take the TrustedType enum instead of a String

ziransun replied:
As discussed, we leave it as String for now and do the clean up at a later stage.

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -173,6 +187,39 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
     return stringValue;
 }
 
+String getTrustedTypeForAttribute(const String& tagName, const String& attributeParameter, const String& elementNamespace, const String& attributeNamespace, String sink)
```
Change this to return `TrustedType` enum rather than a string (you just convert it back anyway)

ziransun replied:
As discussed, we leave it as String for now and do the clean up at a later stage.

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1975,6 +1976,43 @@ const AtomString& Element::getAttributeNS(const AtomString& namespaceURI, const
+        },
+        [](const RefPtr<TrustedScriptURL>& trustedScriptURL) -> ExceptionOr<String> {
+            return trustedScriptURL->toString();
+        }
```
This switch is wrong, if an attribute expects TrustedHTML it shouldn't allow you to pass TrustedScript for example.

You need to do something like below (obviously follow code standards)

```c++
if (attributeType == TrustedXXX) return trustedXXX->toString()
return trustedTypeCompliantString(stringToTrustedType(attributeType), *(element->document().scriptExecutionContext()), trustedXXX->toString(), sink);
```

lukewarlow commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1975,6 +1976,43 @@ const AtomString& Element::getAttributeNS(const AtomString& namespaceURI, const
+static ExceptionOr<String> trustedTypesCompliantAttributeValue(const String& attributeType, const TrustedTypeOrString& value, Element* element, String sink)
+{
+    auto stringValueHolder = WTF::switchOn(value,
+        [&](const String& str) -> ExceptionOr<String> {
```
call this string rather than str

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers-expected.txt:
```diff
@@ -23,109 +116,108 @@ PASS Non-event handler div.accessKey should not be blocked.
+PASS Event handler div.oncontentvisibilityautostatechange should be blocked.
+PASS Event handler div.oncontextmenu should be blocked.
+PASS Event handler div.oncopy should be blocked.
 FAIL Event handler div.oncuechange should be blocked. assert_throws_js: function "_ => element.setAttribute(name, "2+2")" did not throw
```
Interesting that this one doesn't pass?

ziransun replied:
https://searchfox.org/wubkat/source/Source/WebCore/html/HTMLElement.cpp#328

lukewarlow replied:
Hmmm. Fair enough for now. Can you raise a follow up bug to look at that. If one doesn't already exist (I think there is an event handler bug already maybe just note it on that)

ziransun replied:
I found this one https://bugs.webkit.org/show_bug.cgi?id=172985

lukewarlow commented on Source/WebCore/dom/Element.h:
```diff
@@ -35,6 +35,9 @@
 #include "SimulatedClickOptions.h"
+#include "TrustedHTML.h"
+#include "TrustedScript.h"
+#include "TrustedScriptURL.h"
```
A way to get rid of these imports is to make a separate setAttribute and setAttributeNS function which doesn't take the variant (basically the original function) and only export those. The implementation can just call your variant version. That way the webkit legacy doesn't need the concrete types.

lukewarlow reviewed (dismissed):
Would be good to try and get rid of the header imports in the Element.h file.

Edit: Done

ziransun committed:
Implement trusted types integrations with DOM attribute APIs.
https://bugs.webkit.org/show_bug.cgi?id=270436

Reviewed by Ryosuke Niwa.

Implement the spec updates at https://github.com/whatwg/dom/pull/1247

It also removes some expectations in GTK as the results should be
in line with the general expectation file.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-setAttribute-respects-Elements-node-documents-globals-CSP-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/GlobalEventHandlers-onclick-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttributeNS-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/dom/Element.cpp:
(WebCore::trustedTypesCompliantAttributeValue):
(WebCore::Element::validateAttributeIndex const):
(WebCore::Element::toggleAttribute):
(WebCore::Element::setAttribute):
(WebCore::Element::setElementsArrayAttribute):
(WebCore::appendAttributes):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::setAttributeNS):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/dom/TrustedScript.h:
* Source/WebCore/dom/TrustedScriptURL.h:
(WebCore::TrustedScriptURL::toString const): Deleted.
(WebCore::TrustedScriptURL::toJSON const): Deleted.
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::stringToTrustedType):
(WebCore::trustedTypeForAttribute):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
(WebCore::TrustedTypePolicyFactory::getAttributeType const):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMElement.mm:
(-[WKDOMElement setAttribute:value:]):
* Source/WebKitLegacy/mac/DOM/DOMElement.mm:
(-[DOMElement setAttribute:value:]):
(-[DOMElement setAttributeNS:qualifiedName:value:]):

Canonical link: https://commits.webkit.org/278817@main
